### PR TITLE
Improvements to Embedding

### PR DESCRIPTION
**Pull request by wisecornelius**
## Problem

- Embeddings sometimes generate NaN values.

## Proposed changes

1) Better error messages on where NaN values occur.
2) Variable embeddings are now combined through linear layer and not just summed up. The weights / activations of this layer should help to understand the weighting of the feature importances.
3) Continuous feature embeddings are clipped to range [-60k, 60k] s.t. they are in the range of bfloat16.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: neural_lifetimes/models/nets/embedder.py [hunk 1]
```diff
@@ -64,16 +66,31 @@ def build_parameter_dict(self) -> Dict[str, Any]:
...
-            out += F.dropout(self.emb[name](x[name]))
+            disc_emb = F.dropout(self.emb[name](x[name]))
+            assert not disc_emb.isnan().any(), f"Embedding for discrete feature '{name}' contains `NaN` values."
+            combined_emb.append(disc_emb)
+
+        combined_emb = torch.stack(combined_emb, dim=-1)
+        out = self.combine(combined_emb).squeeze()
+        assert not out.isnan().any(), "Combined Feature Embeddings contain `NaN` values."
 
+        out = self.layer_norm(out)  # TODO try removing this again once all features are properly normalized
```
Comment (wisecornelius) [inline neural_lifetimes/models/nets/embedder.py:94]: this helped stabilise with unnormalised features, but it might be worth trying to remove it again later.

Comment (pereplanellTW) [reply neural_lifetimes/models/nets/embedder.py:94]: Shouldn't the normalisation happen at the beginning? If we have different batches with different discrete_value ranges, this normalisation will give the same  `out` normalised values, right?

File: neural_lifetimes/models/nets/embedder.py [hunk 2]
```diff
@@ -64,16 +66,31 @@ def build_parameter_dict(self) -> Dict[str, Any]:
...
     def forward(self, x: Dict[str, torch.Tensor]):
+        combined_emb = []
+
         # batch x num_cont_features
         cf = torch.stack([x[f] for f in self.continuous_features], dim=1)
-        cf[cf.isnan()] = 0
+        cf[cf.isnan()] = 0  # TODO do not do if nan is start token
+
+        cf_emb = F.dropout(F.relu(self.c1(cf)), self.drop_rate, self.training)
+        cf_emb = torch.clip(cf_emb, -65000, 65000)
```
[the remark this answers is not in the document]
Comment (pereplanellTW) [reply neural_lifetimes/models/nets/embedder.py:76]: What would be the right approach in the future?

File: neural_lifetimes/models/nets/embedder.py (hunk 1 above)
Comment (wisecornelius) [reply neural_lifetimes/models/nets/embedder.py:94]: the normalisation of features should happen in the dataloading per feature. This normalises all features together. I think we should remove this.

**Review by pereplanellTW — APPROVED** [review]
